utils/staking-miner/src/main.rs

```diff
@@ -0,0 +1,523 @@
+			if targets.len() == 0 {
+				println!("target = {:?}", (who, stake, targets));
+				return;
+			}
+			let share: u128 = (stake as u128) / (targets.len() as u128);
+			for target in targets {
+				*candidates_and_backing.entry(target.clone()).or_default() += share
+			}
+		});
+
+		let mut candidates_and_backing =
+			candidates_and_backing.into_iter().collect::<Vec<(_, _)>>();
+		candidates_and_backing.sort_by_key(|(_, total_stake)| *total_stake);
+		let winners = candidates_and_backing
+			.into_iter()
+			.rev()
+			.take(desired_targets as usize)
+			.collect::<Vec<_>>();
+		let score = {
+			let min_staker = *winners.last().map(|(_, stake)| stake).unwrap();
+			let sum_stake = winners.iter().fold(0u128, |acc, (_, stake)| acc + stake);
+			let sum_squared = winners.iter().fold(0u128, |acc, (_, stake)| acc + stake);
+			[min_staker, sum_stake, sum_squared]
+		};
+		println!("mined a dpos-like solution with score = {:?}", score);
+		Ok(())
+	})
+}
+
+pub(crate) async fn get_account_info<T: frame_system::Config>(
+	client: &WsClient,
+	who: &T::AccountId,
+	maybe_at: Option<T::Hash>,
+) -> Result<Option<frame_system::AccountInfo<Index, T::AccountData>>, Error> {
+	rpc_helpers::get_storage::<frame_system::AccountInfo<Index, T::AccountData>>(
+		client,
+		params! {
+			sp_core::storage::StorageKey(<frame_system::Account<T>>::hashed_key_for(&who)),
+			maybe_at
+		},
+	)
+	.await
+}
+
+/// Read the signer account's uri from the given `path`.
+async fn read_signer_uri<
+	P: AsRef<Path>,
+	T: frame_system::Config<AccountId = AccountId, Index = Index>,
+>(
+	path: P,
+	client: &WsClient,
+) -> Result<Signer, Error> {
+	let uri = std::fs::read_to_string(path)?;
+
+	// trim any trailing garbage.
+	let uri = uri.trim_end();
+
+	let pair = Pair::from_string(&uri, None)?;
+	let account = T::AccountId::from(pair.public());
+	let _info =
+		get_account_info::<T>(&client, &account, None).await?.ok_or(Error::AccountDoesNotExists)?;
+	log::info!(target: LOG_TARGET, "loaded account {:?}, info: {:?}", &account, _info);
+	Ok(Signer { account, pair, uri: uri.to_string() })
+}
+
+#[tokio::main]
+async fn main() {
+	env_logger::Builder::from_default_env().format_module_path(true).format_level(true).init();
+	let Opt { shared, command } = Opt::from_args();
+	log::debug!(target: LOG_TARGET, "attempting to connect to {:?}", shared.uri);
+
+	let client = loop {
+		let maybe_client = WsClientBuilder::default()
+			.connection_timeout(std::time::Duration::new(20, 0))
+			.max_request_body_size(u32::MAX)
+			.build(&shared.uri)
+			.await;
+		match maybe_client {
+			Ok(client) => break client,
+			Err(why) => {
+				log::warn!(
+					target: LOG_TARGET,
+					"failed to connect to client due to {:?}, retrying soon..",
+					why
+				);
+				std::thread::sleep_ms(2500);
+			}
+		}
+	};
+
+	let chain = rpc_helpers::rpc::<String>(&client, "system_chain", params! {})
+		.await
+		.expect("system_chain infallible; qed.");
```

Comment:
Maybe ok for now, but from my experience these rpcs + websockets can fail at a negligible rate.

Reply:
If we are attempting idempotent RPC calls we could build in some retry logic with exponential backoff

Reply:
For now I only put retry on the initial startup, and for all the rest we assume there's never going to be a failure. If anything fails, I (for now) want to program to crash and restart, and notify us.

Reply:
I'm quite open to improve this, but this way is just much much easier to implement.